feat: new react template based code generation (#23)
asyncapi/go-template: readme.md

- [Overview](#overview)
- [Technical requirements](#technical-requirements)
- [Supported protocols](#supported-protocols)
- [How to use the template](#how-to-use-the-template)
  * [CLI](#cli)
- [Template configuration](#template-configuration)

## Overview

This template generates a Go module that uses [watermill](https://github.com/ThreeDotsLabs/watermill) as the messaging middleware

## Technical requirements

- 1.1.0 =< [Generator](https://github.com/asyncapi/generator/) < 2.0.0,
- Generator specific [requirements](https://github.com/asyncapi/generator/#requirements)

## Supported protocols

Currently this template supports AMQP subscribers

## How to use the template

This template must be used with the AsyncAPI Generator. You can find all available options [here](https://github.com/asyncapi/generator/).

### CLI

> As of this initial commit this template has been tested to generate an AMQP subscriber for the asyncapi.yaml file located in the `test` folder

Reviewer: ```suggestion
This template has been tested to generate an AMQP subscriber for [this asyncapi.yaml file](./test/asyncapi.yaml).
```

how about this? we link to the file and also do not refer to a commit but the template in generate, at state when people read the readme file

defaultContentType: application/json

servers:
  local:
    url: localhost:5672
    protocol: amqp
    security:
      - user-password: []

channels:
  light/measured:
    bindings:
      amqp:
        is: routingKey
        queue:
          name: light/measured
          durable: true
          exclusive: true
          autoDelete: false
          vhost: /
        bindingVersion: 0.2.0
    publish:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: onLightMeasured
      message:
        name: LightMeasured
        payload:
          $id: LightMeasured
          additionalProperties: false
          type: object
          properties:
            id:
              type: integer
              minimum: 0
              description: Id of the streetlight.
            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.
            sentAt:
              type: string
              format: date-time
              description: Date and time when the message was sent.

components:
  securitySchemes:
    user-password:
      type: userPassword
```

> #### Run the following command to generate a golang module

Reviewer 2: ```suggestion
#### Run the following command to generate a Go module
```

```bash
npm install -g @asyncapi/generator
# clone this repository and navigate to this repository
ag /path/to/asyncapi.yaml ./ -o /path/to/generated-code -p moduleName=your-go-module-name
```

#### How to use the generated code

> The above code currently generates a golang module that has a AMQP subscriber.

Reviewer 2: ```suggestion
The above code currently generates a Go module that has a AMQP subscriber.
```

##### Pre-requisites
To run the generated code the following needs to be installed

> 1. go 1.16 +

Reviewer 2: However, you require version 1.15 in `go.mod` file.

> 2. rabbitmq-server

Reviewer 2: Sorry, this is wrong. You should run `docker run -d -p 15672:15672 -p 5672:5672 rabbitmq:3-management` if you wanna also have the UI running at `localhost:15672` for creating queue, publishing message, etc.

##### Running the code

1. Navigate to the path where the code was generated
2. Run the following commands to download the dependencies
```bash
go mod download

> go mod tidy

Author: if i do not run `go mod tidy` i get an error like

```
missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/subscribers.go:6:5: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill-amqp/pkg/amqp (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/handlers.go:9:2: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill/message (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
```

> go mod tidy

Reviewer 2: WDYT about generating the `go.sum` file as well so all checksums are already checked in?

> go mod tidy

Author: since this code is being generated, it would not be an easy task to generate the go.sum
Also from a tooling perspective they way I see it is that this tool is responsible for generating the required code and basic skeleton. From there on it is the responsibility of the consumer to download the required dependencies

> go mod tidy

Reviewer 2: `go.sum` stores the sum (hash) of all dependencies for consistency check. It is not storing dependencies or anything else. 
Generating the `go.sum` should be an atomic process, so it is as easy as just generating a file containing exactly the same content of your local `go.sum` file.

Then, consumers will need only to do `go mod download` for getting an **exact** copy of what you want.

> go mod tidy

Author: hmm I see what you say, however I have 2 follow up questions

1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?
2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

> go mod tidy

Reviewer 2: > 1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?

I feel is the natural process to follow. Otherwise, how would you add a dependency to Go generated code without ensuring that generated code can, let's say, compile?
You could perhaps add a end-to-end functional test where you generate code and run testsuite, linters if you want, etc.
I think better to open a new issue in order to track this effort.

> 2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

Whatever version you add as directive to the `go.mod` file indicates the minimum acceptable Go version by that module. Meaning users should not enter a lower version than that.

> go mod tidy

Author: @smoya then should I not provide the `goVersion` option to the generator and say that users must use go 1.16+ in the readme ?

> go mod tidy

Reviewer 2: `goVersion` option is used **only** to add the constraint to the `go.mod` file. However, that constraint is tied to the generated code, so I think I would +1 to remove it. Maybe if you can validate the `goVersion` option to be 1.16 or greater... but still IMHO is not worth it.

> go mod tidy

Author: this has now been addressed

## Supported protocols

<!--  
Specify what protocols is your code generator supporting. This information should match the information provided in the template configuration under the `supportedProtocols` property. Don't put this section in your readme if your template doesn't generate code.
-->

Currently this template supports AMQP subscribers

## How to use the template

<!--  
Make sure it is easy to try out the template and check what it generates. Instructions for CLI and Docker should be easy to use; just copy/paste to the terminal. In other words, you should always make sure to have ready to use docker-compose set up so the user can quickly check how generated code behaves.
-->

This template must be used with the AsyncAPI Generator. You can find all available options [here](https://github.com/asyncapi/generator/).

### CLI

As of this initial commit this template has been tested to generate an AMQP subscriber for the following asyncapi.yml file

> ```yaml

Reviewer: maybe better put this file in a real file, in test folder, and then your instructions in next sections, to run generator, can point to this file?

channels:
  light/measured:
    bindings:
      amqp:
        is: routingKey
        queue:
          name: light/measured
          durable: true
          exclusive: true
          autoDelete: false
          vhost: /
        bindingVersion: 0.2.0
    publish:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: onLightMeasured
      message:
        name: LightMeasured
        payload:
          $id: LightMeasured
          additionalProperties: false
          type: object
          properties:
            id:
              type: integer
              minimum: 0
              description: Id of the streetlight.
            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.
            sentAt:
              type: string
              format: date-time
              description: Date and time when the message was sent.

components:
  securitySchemes:
    user-password:
      type: userPassword
```

#### Run the following command to generate a Go module

```bash
npm install -g @asyncapi/generator
# clone this repository and navigate to this repository

> ag /path/to/asyncapi.yaml ./ -o /path/to/generated-code -p moduleName=your-go-module-name -p goVersion=1.17

Reviewer: ```suggestion
ag /path/to/asyncapi.yaml @asyncapi/go-watermill-template -o /path/to/generated-code -p moduleName=your-go-module-name -p goVersion=1.17
```

we will release it under official name

```

There are 2 options that can be passed to the generator

> 1. moduleName: name of the go module to be generated

Reviewer: It should be in section like https://github.com/asyncapi/nodejs-template#template-configuration that you already have mentioned in the TOC

2. goVersion: version of Go to be specified in the go.mod file

#### How to use the generated code

The above code currently generates a Go module that has a AMQP subscriber.

##### Pre-requisites
To run the generated code the following needs to be installed

1. go 1.16 +
2. rabbitmq-server OR docker

##### Running the code

1. Navigate to the path where the code was generated
2. Run the following commands to download the dependencies
```bash
go mod download
go mod tidy
```
3. Currently the code does not utilize the server bindings to generate the server URI. It is currently hardcoded to point to a local instance of `rabbitmq`. It is hardcoded as `"amqp://guest:guest@localhost:5672/"` at `<generated-code>/config/server.go`. Change it as per your rabbitmq instance requirements
4. Finally to execute the code run
```bash
go run main.go
```
5. Running local instance of `rabbitmq`, navigate to it using `http://localhost:15672/` with username and password `guest`/ `guest` (These are default rabbitmq credentials). 
   FYI one can start an instance of `rabbitmq` using  `docker` as follow
   ```
   docker run -d -p 15672:15672 -p 5672:5672 rabbitmq:3-management
   ```

> 6. Create a queue as per the async api spec

Reviewer: ```suggestion
6. Create a queue as per the AsyncAPI file
```

Reviewer: honestly would be cool to have at least a link to doc that explains how to do it

Author: the official docs are insufficient, so I am going to provide a link to an article that walks through the UI process.

> 7. Publish a message to the queue as per the async api spec

Reviewer: ```suggestion
7. Publish a message to the queue as per the AsyncAPI file
```

Reviewer: honestly would be cool to have at least a link to doc that explains how to do it, or simple curl, or something

Author: curl is such a good idea I am going to add that
